server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java

```diff
@@ -967,6 +969,13 @@ AllocateUnassignedDecision decideAllocateUnassigned(final ShardRouting shard) {
     }
 
     private static final Comparator<ShardRouting> BY_DESCENDING_SHARD_ID = Comparator.comparing(ShardRouting::shardId).reversed();
+    private static final Comparator<ShardRouting> PRIMARY_FIRST = Comparator.comparing(ShardRouting::primary).reversed();
+
+    private boolean isSegRepEnabled(String index) {
```

Comment:
Shard Balancing algo shouldn't be aware of replication. Have method inside the IndexMetadata which return the mode PRIMARY_FIRST or not and that could check whether it is segrep enabled index or not.

Reply:
Thank you @shwetathareja for the feedback and review. I will update this as you suggested.

This is an improvement over #6017 where primary shard balancing factor was introduced. It would be great if you can review and share your thoughts on issue https://github.com/opensearch-project/OpenSearch/issues/5240

Reply:
My initial thoughts:

Have you looked into adding an AllocationConstraint similar to [isIndexShardsPerNodeBreached](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationConstraints.java#L79) which essentially adds constraint to ensure each node shouldn't have more than avg. no.of shards per index even if a node has overall less shards compared to other nodes. This helps in preventing un-necessary rebalances later. More details are here - https://github.com/opensearch-project/OpenSearch/issues/487

I feel you can simply add a constraint for segrep indices to ensure a node shouldn't have more than avg. no. of expected primaries for a index. Lets say it is 5 node cluster and segrep index has 5 primaries + 2 replicas. So no node should have more than 1 primary. This wouldn't require coming up with magic value for weight factor. Later, we can evaluate if we want to extend this to balancing sum of primaries across all indices across all nodes as well. e.g. 1 segrep index has 3 primaries while other has 2 primaries and it is 5 node cluster, then each node shouldn't have more than 1 primary across indices as well.

Reply:
Thank you @shwetathareja for the feedback. This is a great idea! 

Yes, I think having a constraint (here, avg primary shard count/node) supplements the existing weight function and allows more finer control. I think it should also fixes the current problems of uneven primary shard distribution from failover and rebalacing which works only on individual indices. 

The [existing balancer](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java#L181) performs balancing at single index level i.e. balnce shards of one index across all nodes, repeat for others. We can introduce a new balancing logic which works across all indices and across all nodes. A simple algorithm can be:

####Constants
Allocation constraint weight when node contains more than avg primary shards, lets say `Wa = 1000`
Threshold, say `t = 2 * Wa`. When node weight differ more than threshold, it is considered unbalanced. We use twice of `Wa` to avoid repeatitive shard movements.
#### Steps
1. Calculate weight on all nodes using weight with constraint above. `E.g. 2 nodes, say N1, N2 and 5 shards. N1 contains 2 shard, N2 contains 3. With avg shard count/ node as 2.5, N1 will have `0` weight while N2 have `1000`. 
2. Sort nodes based on constraint weight.
3. while(true)
      - Assign light node and heavy node
      - If weight diff > threshold, move *random* primary from heavy -> light node
         - update weight of both nodes
         - sort and continue to move more shards
      - else break


### Approach 1.
Additional round of balancing based on avg primary shard/node constraint. First balance on individual indices (existing functionality) and then balance across all nodes. 
Pros.
1. Have better per index shard distribution 
Cons. 
1. Additional relocations from second round.
2. Problems of
     - Node level balancing undoing index level balancing operations. Example is
        - 2 Nodes, say N1, N2, 3 indices (I1, I2, I3) with total 4 primary shards. 
        - I1 has I1P1, I1P2 as primary. I2 has I2P1, I3 has I3P1
        - N1 -> I1P1, I1P2 while N2 contains I2P1, I3P1. Index level balancing moves I1P1 from N1 to N2. Node level balancing moves it back
     - Endless loop of balancing both undoing actions from previous.

### Approach 2. 
Balancer performs balancing across all indices and nodes once using algorithm above for segrep specific shards. 

Pros. 
1. Minimal shard relocations.

Cons.
1. Chances of skewed per index shard distribution